File: 2D coordinate system transformation/conformal/conformal_2D.py
import pandas as pd
import numpy as np

# Read the Excel file
df = pd.read_excel('points_coordinate_system_1.xlsx')
df2 = pd.read_excel('control_points.xlsx')
df3 = pd.read_excel('old_coordinates.xlsx')
df4 = pd.read_excel('check1.xlsx')
df5 = pd.read_excel('check2.xlsx')

# Convert DataFrame to a list of lists (each row as a list)
points_coordinate_system_1 = df.values.tolist()
control_points = df2.values.tolist()
old_coordinates = df3.values.tolist()
chekc1 = df4.values.tolist()
chekc2 = df5.values.tolist()

# equations and unknowns
n = len(control_points) * 2
u = 4

# creating matrix of observations for least square algorithm
L = np.zeros((n, 1))

c = 0
for i in control_points:
    L[c][0] = i[0]
    c += 1
    L[c][0] = i[1]
    c += 1

# creating matrix A for least square algorithm
def create_A(arr, n, u):
    A = np.zeros((n, u))
    c = 0
    for i in arr:
        A[c][0] = i[0]
        A[c][1] = i[1]
        A[c][2] = 1
        A[c][3] = 0

        c += 1

        A[c][0] = i[1]
        A[c][1] = -i[0]
        A[c][2] = 0
        A[c][3] = 1

        c += 1
    return A

A = create_A(points_coordinate_system_1, n, u)

# calculating unknowns for 2D conformal transformation
X_CAP = np.linalg.inv(A.T @ A) @ A.T @ L

a = X_CAP[0][0]
b = X_CAP[1][0]
X0 = X_CAP[2][0]
Y0 = X_CAP[3][0]

# ----------------------------- RMSE CALCULATION -----------------------------
n = len(chekc1) * 2

A = create_A(chekc1, n, u)

new_check_coordinates = A @ X_CAP
new_check_coordinates = np.reshape(new_check_coordinates, newshape=(int(np.ceil(len(new_check_coordinates) / 2)), 2))

residuals = chekc2 - new_check_coordinates
residual_x = residuals[:, 0]
residual_y = residuals[:, 1]

sum_e = 0
for i in range(len(residual_x)):
    e = np.sqrt(residual_x[i] **2 + residual_y[i]**2)
    sum_e += (e**2)

rmse = np.sqrt(sum_e / len(residual_x))

print(f"Total RMSE: {rmse:.10f}")

# calculating scale
lambda_ = np.sqrt(a**2 + b**2)

# calculating kappa angle
K = np.rad2deg(np.arctan2(b, a))

# moving our coordinates to them new coordinates system
n = len(old_coordinates) * 2
A = create_A(old_coordinates, n, u)

new_coordinates = A @ X_CAP
new_coordinates = np.reshape(new_coordinates, newshape=(int(np.ceil(len(new_coordinates) / 2)), 2))

# creating csv file of our results
df4 = pd.DataFrame(new_coordinates, columns=["X_new", "Y_new"])
df4.to_csv(f"new_coordinates.csv", index=False)

File: 2D coordinate system transformation/conformal/conformal_2D.m
clc;clear;close all;

% Read the Excel file
points_coordinate_system_1 = xlsread('points_coordinate_system_1.xlsx');
control_points = xlsread('control_points.xlsx');
old_coordinates = xlsread('old_coordinates.xlsx');
check1 = xlsread('check1.xlsx');
check2 = xlsread('check2.xlsx');

% Equations and unknowns
n = size(control_points, 1) * 2;
u = 4;

% Creating matrix of observations for least square algorithm
L = zeros(n, 1);

c = 1; % MATLAB indexing starts at 1
for i = 1:size(control_points, 1)
    L(c, 1) = control_points(i, 1);
    c = c + 1;
    L(c, 1) = control_points(i, 2);
    c = c + 1;
end

% Create matrix A
A = create_A(points_coordinate_system_1, n, u);

% Calculating unknowns for 2D conformal transformation
X_CAP = inv(A' * A) * A' * L;

% Compute new coordinates of control points using existing A and X_CAP
n3 = size(check1, 1) * 2;
A3 = create_A(check1, n3 ,u);
new_check = A3 * X_CAP;
new_check = reshape(new_check, 2, [])';

% Calculate residuals between real and transformed control points
residuals = check2 - new_check;

residual_x = residuals(:, 1);
residual_y = residuals(:, 2);

sum_e = 0;
for i = 1:length(residual_x)
    e = sqrt(residual_x(i)^2 + residual_y(i)^2);
    sum_e = sum_e + e^2;
end

rmse = sqrt(sum_e / length(residual_x));

fprintf('Total RMSE: %.10f\n', rmse);

a = X_CAP(1, 1);
b = X_CAP(2, 1);
X0 = X_CAP(3, 1);
Y0 = X_CAP(4, 1);

% Calculating scale
lambda = sqrt(a^2 + b^2);

% Calculating kappa angle
K = rad2deg(atan2(b, a));

% Moving our coordinates to the new coordinate system
n = size(old_coordinates, 1) * 2;
A = create_A(old_coordinates, n, u);
new_coordinates = A * X_CAP;
[s, ~] = size(new_coordinates);

final_coordinates = zeros(s/2, 2);

c = 1;
for i = 1:(s/2)
    final_coordinates(i, 1) = new_coordinates(c);
    c = c + 1;
    final_coordinates(i, 2) = new_coordinates(c);
    c = c + 1;
end

% Creating CSV file of our results
csvwrite('new_coordinates_matlab.csv', final_coordinates);

% Function to create matrix A for least square algorithm
function A = create_A(arr, n, u)
    A = zeros(n, u);
    c = 1; % MATLAB indexing starts at 1
    for i = 1:size(arr, 1)
        A(c, 1) = arr(i, 1);
        A(c, 2) = arr(i, 2);
        A(c, 3) = 1;
        A(c, 4) = 0;
        
        c = c + 1;
        
        A(c, 1) = arr(i, 2);
        A(c, 2) = -arr(i, 1);
        A(c, 3) = 0;
        A(c, 4) = 1;
        
        c = c + 1;
    end
end